### DISCUSSION TITLE
NO-COPY examples running ping app under Linux core and echo application under baremetal?

### DISCUSSION THREAD

**Discussion by ravijj** [OpenAMP/open-amp#294] [Q&A]
Hi All,

We are currently working on Openamp integration in the Xilinx Zedboard.
One core cortex A9 running the Linux and another cortex A9 core running the bare-metal application.

Our aim is to communicate between these two core using the OpenAMP framework. 
We have verified the echo example application one is running under the Linux and another one running as a bare-metal application.

Now, we are looking into no-copy communication between the Linux and bare-metal cores.
The existing no-copy example is demonstrating the no-copy communication running in the two Linux applications ping and echo.

Is there any example that demonstrates the no-copy application between Linux and bare-metal?

Your help is greatly appreciated.

Thanks and Regards,
Ravikiran J J

**Comment by arnopo** [comment]
Hello @ravijj

If the echo exemple work in your environement I would have expected  zero_copy exemple work too. could you share details on issue you are facing?

That's said if you have the echo sample running in your environment, moving to the zero copy should not be a big stuff.
Just compare both exemples : [echo](https://github.com/OpenAMP/open-amp/tree/master/apps/examples/echo) & [nocopy_echo](https://github.com/OpenAMP/open-amp/tree/master/apps/examples/nocopy_echo)
Delta is mainly around the rpmsg API use and the way of managing the buffers in the application

Notice that you can do this in 2 steps, for instance keep normal echo test on Linux side and move to the zero-copy on bare metal, then move lInux appli on the zerocopy API

**Reply by ravijj** [reply]
Hello Arnopo,

Thank you so much for the reply.

Okay there is one clarification I need to give it here.
The Slave baremetal application is from this open-amp source  **echo** : **"apps/examples/echo/rpmsg-echo.c"**
**The Linux Master/Ping application I used, it is not from this open-amp source **echo**, It is from the **meta-openamp** layer build **echo_test.c** application.**
One cortex-a9 core is running Linux + echo_test application and other cortex-a9 core is running baremetal  this open-amp rpmsg_echo.c application and these two applications are communicating fine each other in the Xilinx Zynq Zedboard.

NO-COPY application: 
What we are trying now is, we have compiled the  baremetal application **nocopy_echo**: **"apps/examples/echo/rpmsg-nocopy-echo.c"** for the slave core.
We didn't find the master Linux nocopy_echo_test application. in the meta-openamp layer.

1. Where we can find the master Linux  nocopy echo test application source code?
2. Is it possible to compile the **echo:** **"apps/examples/echo/rpmsg-echo.c"** or **nocopy_echo:** **"apps/examples/echo/rpmsg-nocopy-echo.c"** for Zynq7 Linux master and it can communicate with  the slave core running  baremetal  **echo or nocopy_echo application.**

Thank you for looking into this matter in advance.

Thanks and Regards,
Ravikiran J J

**Reply by arnopo** [reply]
> Hello Arnopo,
> 
> Thank you so much for the reply.
> 
> Okay there is one clarification I need to give it here.
> The Slave baremetal application is from this open-amp source **echo** : **"apps/examples/echo/rpmsg-echo.c"**
> **The Linux Master/Ping application I used, it is not from this open-amp source **echo**, It is from the **meta-openamp** layer build **echo_test.c** application.**
> One cortex-a9 core is running Linux + echo_test application and other cortex-a9 core is running baremetal this open-amp rpmsg_echo.c application and these two applications are communicating fine each other in the Xilinx Zynq Zedboard.
> 
> NO-COPY application:
> What we are trying now is, we have compiled the baremetal application **nocopy_echo**: **"apps/examples/echo/rpmsg-nocopy-echo.c"** for the slave core.
> We didn't find the master Linux nocopy_echo_test application. in the meta-openamp layer.

This repo is independent of the library, it was maintained by Xilinx, but does not seem to have been updated for some time.
@edmooring do you have any info on this repo?

> 
> 1. Where we can find the master Linux  nocopy echo test application source code?

It does not exist but you should be able to reuse  https://github.com/OpenAMP/open-amp/blob/master/apps/examples/nocopy_echo/rpmsg-nocopy-ping.c to construct it

> 2. Is it possible to compile the **echo:** **"apps/examples/echo/rpmsg-echo.c"** or **nocopy_echo:** **"apps/examples/echo/rpmsg-nocopy-echo.c"** for Zynq7 Linux master and it can communicate with  the slave core running  baremetal  **echo or nocopy_echo application.**

No dependence each side can choice to implement the copy or no_copy method

> 
> Thank you for looking into this matter in advance.
> 
> Thanks and Regards,
> Ravikiran J J

**Reply by edmooring** [reply]
The meta-openamp repo is still maintained by Xilinx, but isn't updated regularly. 
You won't be able to modify the examples to do what you want because the no-copy functions are unavailable in the Linux kernel rpmsg implementation, which is what these examples use.

You could do this by switching to the Linux user-space rpmsg examples. This requires modifications to the kernel device tree.

**Reply by ravijj** [reply]
Hi Arnopo and Edmooring,

Is it possible to compile the no-copy master application nocopy_echo: "apps/examples/echo/rpmsg-nocopy-ping.c" to Zynq7 Linux userspace?
So that we can make it to communicate with the Baremetal nocopy_echo: "apps/examples/echo/rpmsg-nocopy-echo.c" application running on the other core.

Here we are trying to make the open-amp no-copy example working between core to core instead of between the two Linux processes.

Thanks and Regards,
Ravikiran J J

**Comment by ravijj** [comment]
Hi Arnopo and Edmooring,

Is it possible to compile the no-copy master application **nocopy_echo: "apps/examples/echo/rpmsg-nocopy-ping.c" to Zynq7 Linux userspace?** 
So that we can make it to communicate with the Baremetal nocopy_echo: "apps/examples/echo/rpmsg-nocopy-echo.c" application running on the other core.

Here we are trying to make the open-amp no-copy example working **between core to core** instead of between the two Linux processes.

Thanks and Regards,
Ravikiran J J
